File: app/src/main/java/acr/browser/lightning/database/downloads/DownloadsDatabase.java
package acr.browser.lightning.database.downloads;

import android.app.Application;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.DatabaseUtils;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import com.anthonycr.bonsai.Completable;
import com.anthonycr.bonsai.CompletableAction;
import com.anthonycr.bonsai.CompletableSubscriber;
import com.anthonycr.bonsai.Single;
import com.anthonycr.bonsai.SingleAction;
import com.anthonycr.bonsai.SingleSubscriber;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import acr.browser.lightning.R;

/**
 * The disk backed download database.
 * See {@link DownloadsModel} for method
 * documentation.
 */
@Singleton
public class DownloadsDatabase extends SQLiteOpenHelper implements DownloadsModel {

    private static final String TAG = "DownloadsDatabase";

    // Database Version
    private static final int DATABASE_VERSION = 1;

    // Database Name
    private static final String DATABASE_NAME = "downloadManager";

    // HistoryItems table name
    private static final String TABLE_DOWNLOADS = "download";

    // HistoryItems Table Columns names
    private static final String KEY_ID = "id";
    private static final String KEY_URL = "url";
    private static final String KEY_TITLE = "title";
    private static final String KEY_SIZE = "size";

    @NonNull private final String DEFAULT_DOWNLOADS_TITLE;

    @Nullable private SQLiteDatabase mDatabase;

    @Inject
    public DownloadsDatabase(@NonNull Application application) {
        super(application, DATABASE_NAME, null, DATABASE_VERSION);
        DEFAULT_DOWNLOADS_TITLE = application.getString(R.string.untitled);
    }

    /**
     * Lazily initializes the database
     * field when called.
     *
     * @return a non null writable database.
     */
    @WorkerThread
    @NonNull
    private synchronized SQLiteDatabase lazyDatabase() {
        if (mDatabase == null || !mDatabase.isOpen()) {
            mDatabase = getWritableDatabase();
        }

        return mDatabase;
    }

    // Creating Tables
    @Override
    public void onCreate(@NonNull SQLiteDatabase db) {
        String CREATE_BOOKMARK_TABLE = "CREATE TABLE " +
                DatabaseUtils.sqlEscapeString(TABLE_DOWNLOADS) + '(' +
                DatabaseUtils.sqlEscapeString(KEY_ID) + " INTEGER PRIMARY KEY," +
                DatabaseUtils.sqlEscapeString(KEY_URL) + " TEXT," +
                DatabaseUtils.sqlEscapeString(KEY_TITLE) + " TEXT," +
                DatabaseUtils.sqlEscapeString(KEY_SIZE) + " TEXT" + ')';
        db.execSQL(CREATE_BOOKMARK_TABLE);
    }

    // Upgrading database
    @Override
    public void onUpgrade(@NonNull SQLiteDatabase db, int oldVersion, int newVersion) {
        // Drop older table if it exists
        db.execSQL("DROP TABLE IF EXISTS " + DatabaseUtils.sqlEscapeString(TABLE_DOWNLOADS));
        // Create tables again
        onCreate(db);
    }

    @NonNull
    private static ContentValues bindBookmarkToContentValues(@NonNull DownloadItem downloadItem) {
        ContentValues contentValues = new ContentValues(3);
        contentValues.put(KEY_TITLE, downloadItem.getTitle());
        contentValues.put(KEY_URL, downloadItem.getUrl());
        contentValues.put(KEY_SIZE, downloadItem.getContentSize());

        return contentValues;
    }

    @NonNull
    private static DownloadItem bindCursorToDownloadItem(@NonNull Cursor cursor) {
        DownloadItem download = new DownloadItem();

        download.setUrl(cursor.getString(cursor.getColumnIndex(KEY_URL)));
        download.setTitle(cursor.getString(cursor.getColumnIndex(KEY_TITLE)));
        download.setContentSize(cursor.getString(cursor.getColumnIndex(KEY_SIZE)));

        return download;
    }

    @NonNull
    private static List<DownloadItem> bindCursorToDownloadItemList(@NonNull Cursor cursor) {
        List<DownloadItem> downloads = new ArrayList<>();

        while (cursor.moveToNext()) {
            downloads.add(bindCursorToDownloadItem(cursor));
        }

        cursor.close();

        return downloads;
    }

    @NonNull
    @Override
    public Single<DownloadItem> findDownloadForUrl(@NonNull final String url) {
        return Single.create(new SingleAction<DownloadItem>() {
            @Override
            public void onSubscribe(@NonNull SingleSubscriber<DownloadItem> subscriber) {
                Cursor cursor = lazyDatabase().query(TABLE_DOWNLOADS, null, KEY_URL + "=?", new String[]{url}, null, null, "1");

                if (cursor.moveToFirst()) {
                    subscriber.onItem(bindCursorToDownloadItem(cursor));
                } else {
                    subscriber.onItem(null);
                }

                cursor.close();
                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Single<Boolean> isDownload(@NonNull final String url) {
        return Single.create(new SingleAction<Boolean>() {
            @Override
            public void onSubscribe(@NonNull SingleSubscriber<Boolean> subscriber) {
                Cursor cursor = lazyDatabase().query(TABLE_DOWNLOADS, null, KEY_URL + "=?", new String[]{url}, null, null, null, "1");

                subscriber.onItem(cursor.moveToFirst());

                cursor.close();
                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Single<Boolean> addDownloadIfNotExists(@NonNull final DownloadItem item) {
        return Single.create(new SingleAction<Boolean>() {
            @Override
            public void onSubscribe(@NonNull SingleSubscriber<Boolean> subscriber) {
                Cursor cursor = lazyDatabase().query(TABLE_DOWNLOADS, null, KEY_URL + "=?", new String[]{item.getUrl()}, null, null, "1");

                if (cursor.moveToFirst()) {
                    cursor.close();
                    subscriber.onItem(false);
                    subscriber.onComplete();
                    return;
                }

                cursor.close();

                long id = lazyDatabase().insert(TABLE_DOWNLOADS, null, bindBookmarkToContentValues(item));

                subscriber.onItem(id != -1);
                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Completable addDownloadsList(@NonNull final List<DownloadItem> bookmarkItems) {
        return Completable.create(new CompletableAction() {
            @Override
            public void onSubscribe(@NonNull CompletableSubscriber subscriber) {
                lazyDatabase().beginTransaction();

                for (DownloadItem item : bookmarkItems) {
                    addDownloadIfNotExists(item).subscribe();
                }

                lazyDatabase().setTransactionSuccessful();
                lazyDatabase().endTransaction();

                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Single<Boolean> deleteDownload(@NonNull final String url) {
        return Single.create(new SingleAction<Boolean>() {
            @Override
            public void onSubscribe(@NonNull SingleSubscriber<Boolean> subscriber) {
                int rows = lazyDatabase().delete(TABLE_DOWNLOADS, KEY_URL + "=?", new String[]{url});

                subscriber.onItem(rows > 0);
                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Completable deleteAllDownloads() {
        return Completable.create(new CompletableAction() {
            @Override
            public void onSubscribe(@NonNull CompletableSubscriber subscriber) {
                lazyDatabase().delete(TABLE_DOWNLOADS, null, null);

                subscriber.onComplete();
            }
        });
    }

    @NonNull
    @Override
    public Single<List<DownloadItem>> getAllDownloads() {
        return Single.create(new SingleAction<List<DownloadItem>>() {
            @Override
            public void onSubscribe(@NonNull SingleSubscriber<List<DownloadItem>> subscriber) {
                Cursor cursor = lazyDatabase().query(TABLE_DOWNLOADS, null, null, null, null, null, null);

                subscriber.onItem(bindCursorToDownloadItemList(cursor));
                subscriber.onComplete();

                cursor.close();
            }
        });
    }

    @Override
    public long count() {
        return DatabaseUtils.queryNumEntries(lazyDatabase(), TABLE_DOWNLOADS);
    }

}

File: app/src/main/java/acr/browser/lightning/database/downloads/DownloadsDatabase.kt
package acr.browser.lightning.database.downloads

import acr.browser.lightning.browser.di.DatabaseScheduler
import acr.browser.lightning.database.databaseDelegate
import acr.browser.lightning.extensions.firstOrNullMap
import acr.browser.lightning.extensions.useMap
import android.annotation.SuppressLint
import android.app.Application
import android.content.ContentValues
import android.database.Cursor
import android.database.DatabaseUtils
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The disk backed download database. See [DownloadsRepository] for function documentation.
 */
@SuppressLint("Range")
@Singleton
class DownloadsDatabase @Inject constructor(
    application: Application,
    @DatabaseScheduler
    private val databaseDispatcher: CoroutineDispatcher,
) : SQLiteOpenHelper(application, DATABASE_NAME, null, DATABASE_VERSION), DownloadsRepository {

    private val database: SQLiteDatabase by databaseDelegate()

    // Creating Tables
    override fun onCreate(db: SQLiteDatabase) {
        val createDownloadsTable =
            "CREATE TABLE ${DatabaseUtils.sqlEscapeString(TABLE_DOWNLOADS)}(" +
                "${DatabaseUtils.sqlEscapeString(KEY_ID)} INTEGER PRIMARY KEY," +
                "${DatabaseUtils.sqlEscapeString(KEY_URL)} TEXT," +
                "${DatabaseUtils.sqlEscapeString(KEY_TITLE)} TEXT," +
                "${DatabaseUtils.sqlEscapeString(KEY_SIZE)} TEXT" +
                ')'
        db.execSQL(createDownloadsTable)
    }

    // Upgrading database
    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // Drop older table if it exists
        db.execSQL("DROP TABLE IF EXISTS ${DatabaseUtils.sqlEscapeString(TABLE_DOWNLOADS)}")
        // Create tables again
        onCreate(db)
    }

    override suspend fun findDownloadForUrl(
        url: String
    ): DownloadEntry? = withContext(databaseDispatcher) {
        database.query(
            TABLE_DOWNLOADS,
            null,
            "$KEY_URL=?",
            arrayOf(url),
            null,
            null,
            "1"
        ).firstOrNullMap { it.bindToDownloadItem() }
    }

    override suspend fun isDownload(url: String): Boolean = withContext(databaseDispatcher) {
        database.query(
            TABLE_DOWNLOADS,
            null,
            "$KEY_URL=?",
            arrayOf(url),
            null,
            null,
            null,
            "1"
        ).use { it.moveToFirst() }
    }

    override suspend fun addDownloadIfNotExists(
        entry: DownloadEntry
    ): Boolean = withContext(databaseDispatcher) {
        database.query(
            TABLE_DOWNLOADS,
            null,
            "$KEY_URL=?",
            arrayOf(entry.url),
            null,
            null,
            "1"
        ).use {
            if (it.moveToFirst()) {
                return@withContext false
            }
        }

        val id = database.insert(TABLE_DOWNLOADS, null, entry.toContentValues())

        return@withContext id != -1L
    }

    override suspend fun addDownloadsList(
        downloadEntries: List<DownloadEntry>
    ): Unit = withContext(databaseDispatcher) {
        database.apply {
            beginTransaction()
            setTransactionSuccessful()

            for (item in downloadEntries) {
                addDownloadIfNotExists(item)
            }

            endTransaction()
        }
    }

    override suspend fun deleteDownload(url: String): Boolean = withContext(databaseDispatcher) {
        database.delete(TABLE_DOWNLOADS, "$KEY_URL=?", arrayOf(url)) > 0
    }

    override suspend fun deleteAllDownloads(): Unit = withContext(databaseDispatcher) {
        database.run {
            delete(TABLE_DOWNLOADS, null, null)
            close()
        }
    }

    override suspend fun getAllDownloads(): List<DownloadEntry> = withContext(databaseDispatcher) {
        database.query(
            TABLE_DOWNLOADS,
            null,
            null,
            null,
            null,
            null,
            "$KEY_ID DESC"
        ).useMap { it.bindToDownloadItem() }
    }

    override suspend fun count(): Long = withContext(databaseDispatcher) {
        DatabaseUtils.queryNumEntries(database, TABLE_DOWNLOADS)
    }

    /**
     * Maps the fields of [DownloadEntry] to [ContentValues].
     */
    private fun DownloadEntry.toContentValues() = ContentValues(3).apply {
        put(KEY_TITLE, title)
        put(KEY_URL, url)
        put(KEY_SIZE, contentSize)
    }

    /**
     * Binds a [Cursor] to a single [DownloadEntry].
     */
    private fun Cursor.bindToDownloadItem() = DownloadEntry(
        url = getString(getColumnIndex(KEY_URL)),
        title = getString(getColumnIndex(KEY_TITLE)),
        contentSize = getString(getColumnIndex(KEY_SIZE))
    )

    companion object {

        // Database version
        private const val DATABASE_VERSION = 1

        // Database name
        private const val DATABASE_NAME = "downloadManager"

        // DownloadItem table name
        private const val TABLE_DOWNLOADS = "download"

        // DownloadItem table columns names
        private const val KEY_ID = "id"
        private const val KEY_URL = "url"
        private const val KEY_TITLE = "title"
        private const val KEY_SIZE = "size"

    }

}
